code/modules/projectiles/guns/energy/cell_loaded_vr/multi_cannon_cells.dm

```diff
@@ -0,0 +1,102 @@
+/obj/item/ammo_casing/macrobattery
+	caliber = "macrobat"
+	name = "macrobattery"
+	icon = 'icons/obj/ammo_vr.dmi'
+	icon_state = "macrobat_wtf"
+	desc = "A large nanite fabricator for a medigun. Powered by a mix of precursor and modern tech, this fancy device recharges without an outside power source."
+
+	projectile_type = /obj/item/projectile/beam/chain_lightning //why the hell not
+
+	var/bat_colour = "#ff33cc"
+	var/charge
+	var/max_charge = 10
+	var/ticks = 1
+	var/ticks_to_charge = 5 //10 secs per shot charged
+
+/obj/item/ammo_casing/macrobattery/Initialize(mapload, ...)
+	. = ..()
+	START_PROCESSING(SSobj, src)
+	charge = max_charge
+
+/obj/item/ammo_casing/macrobattery/Destroy()
+	STOP_PROCESSING(SSobj, src)
+	. = ..()
+	
+/obj/item/ammo_casing/macrobattery/process()
+	ticks++
+	if(ticks%ticks_to_charge == 0)
+		recharge()
+		if(charge >= max_charge)
+			return PROCESS_KILL
+
+/obj/item/ammo_casing/macrobattery/expend()
+	if(charge)
+		charge --
+		ticks = 1 //so we have to start over on the charge time.
+		START_PROCESSING(SSobj, src)
+		update_icon()
+		. = BB
+		//alright, the below seems jank. it IS jank, but for whatever reason I can't reuse BB. big bad
+		BB = null
+		BB = new projectile_type
+		return 
+	else
+		BB = null
+		return null
```

Comment:
If, for whatever reason (trying to fire with the battery uncharged?), you get here with 0 charge, it will not add itself to processing and start recharging. It's also not great that it tries to add itself to SSobj every time it's fired, but there's a safety check in that macro that will avoid readding it every time, though it will still have to perform said check.

Reply:
Yeah, I can't see that check being resource intensive enough for it to ever wind up being an issue, especially with the inherent ratelimiting clickdelays give.
and I can't see a way for the cells to ever get fired at zero charge *without* them having been already put in the processing subsystem. Maybe varediting them to zero when they're full, and not in the subsystem would do it, but that's a pretty niche case.

Reply:
Or starting empty.

Reply:
Not possible, thankfully! it's charge value is set to max_charge right off the bat in Initialize.

Reply:
Ye, but there could be subtypes in the future that start empty! Plenty of things in the game are like that, like cells and such. I could definitely see if these were producible in game, they'd startp empty like cells ~~are supposed to, and sometimes do~~

Reply:
That's fair, though I think the fluff Natje's got for them with the Precursor stuff would effectively make them impossible to have empty, though I'll change it either way.

Any suggestions for fixing this one? having START_PROCESSING at the start of expend seems kinda sloppy. I could also probably throw it in Initialize.